storm-core/src/jvm/org/apache/storm/daemon/metrics/BuiltinMetricsUtil.java

```diff
@@ -0,0 +1,77 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.storm.daemon.metrics;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.storm.Config;
+import org.apache.storm.metric.api.IMetric;
+import org.apache.storm.metric.api.IStatefulObject;
+import org.apache.storm.metric.api.StateMetric;
+import org.apache.storm.stats.BoltExecutorStats;
+import org.apache.storm.stats.SpoutExecutorStats;
+import org.apache.storm.stats.StatsUtil;
+import org.apache.storm.task.TopologyContext;
+
+public class BuiltinMetricsUtil {
+    public static BuiltinMetrics mkData(String type, Object stats) {
+        if (StatsUtil.SPOUT.equals(type)) {
+            return new BuiltinSpoutMetrics((SpoutExecutorStats) stats);
+        }
+        return new BuiltinBoltMetrics((BoltExecutorStats) stats);
+    }
+
+    public static void registerIconnectionServerMetric(Object server, Map stormConf, TopologyContext context) {
+        if (server instanceof IStatefulObject) {
+            registerMetric("__recv-iconnection", new StateMetric((IStatefulObject) server), stormConf, context);
+        }
+    }
+
+    public static void registerIconnectionClientMetrics(final Map nodePort2socket, Map stormConf, TopologyContext context) {
+        IMetric metric = new IMetric() {
+            @Override
+            public Object getValueAndReset() {
+                Map<Object, Object> ret = new HashMap<>();
+                for (Object o : nodePort2socket.entrySet()) {
```

Comment:
No, it has to be cast to Map.Entry inside the `for` loop code because it's actually a clojure map which has no generic definition. Java compiler will complain if I replace `Object o` to `Map.Entry o`